### ISSUE TITLE
[Design] Support for GraphQL

### ISSUE THREAD

**Issue by serras** [higherkindness/mu-haskell#46]
The goal of this ticket is to design the new adapter from Mu to [GraphQL](https://graphql.org/learn/). Some of the design is based on [Sangria](https://sangria-graphql.org/getting-started/#define-a-graphql-schema), so you may want to check that first.

## The ingredients

GraphQL is made, as most of the protocols supported by Mu, of four parts:

* [A way to define schemas](https://graphql.org/learn/schema/): the main difference from other schema languages is that fields in GraphQL may have arguments;
* [A way to call the operations in those schemas](https://graphql.org/learn/queries/): GraphQL defines "queries", "mutations", and "subscriptions", but they are all quite similar. A very important property of GraphQL is that you must ask which of the fields of each record you want to be returned;
* A way to execute those calls (the *client*);
* Something which understand these calls and returns the corresponding values (the *server*).
 
## Schemas and operations

`mu-schema` already offers a language to describe schemas which is very close to GraphQL's. The only missing piece is the [optional arguments that GraphQL fields may have](https://graphql.org/learn/schema/#arguments). Luckily, we have support for annotations in Mu, which we currently use only to [indicate the field ids in Protocol Buffers](https://github.com/higherkindness/mu-haskell/blob/master/adapter/protobuf/test/ProtoBuf.hs#L40).

## Data

There are two forms of data in use by GraphQL:

1. When you ask for data (or perform a mutation), you have to specify the [fields you want to obtain](https://graphql.org/learn/queries/#fields).
2. The data gathered (or the result of the mutation) is returned as a simple JSON.

How can we represent these in Mu? Since the introduction of [higher-kinded data](https://github.com/higherkindness/mu-haskell/pull/41) we have a "layer" in the terms which can be defined by a `Functor`. We use it, for example, in Protocol Buffers, because the spec mandates all fields to be optional; hence we choose `Maybe` as the "wrapper functor". We can use the same idea to represent the gathered data (item 2 in the list), since data may or may not be present.

The case of (1) is a bit more complex, since you have three options: "I want this field", "I don't want this field", or "I want a subset of inner fields of this field". We can represent this by a new `Functor`:

```haskell
data Wanted a = Wanted | NotWanted | Subset a
```

## Execution

With these pieces we can represent a "server" for a GraphQL query as a function with type:

```haskell
execute :: Query Wanted -> Query Maybe
```

with the additional invariant that every field which was `Wanted` appears as `Just` in the result.

However, this requires too much work from the programmer. Instead, we can take [some ideas from Sangria](https://sangria-graphql.org/getting-started/#define-a-graphql-schema) and specify the "resolver" of each field separately. These *resolvers*, in Sangria, take as input the parent value, so a field may just request an actual field from the value from the parent. In summary, we want to define a data structure which gathers all the resolvers of a schema. This can be done via GADTs (here, `TypeDef` is the kind which represents each of the types in the schema).

```haskell
-- one resolver per type in the schema
type SchemaResolver = NP (TypeResolver sch)

data TypeResolver (sch :: Schema) ty where
  -- one resolver per field
  RRecord :: NP (FieldResolver sch name) args
          -> TypeResolver sch ('DRecord name args)
  ... REnum, RSimple ...

data FieldResolver (sch :: Schema) tyname f where
  -- obtain a value given the parent
  RField :: (Term sch (sch :/: tyname) -> FieldValue sch v)
         -> FieldResolver sch tyname ('FieldDef fname v)
```

From this, it should be possible to define a complete query:

```haskell
resolvers :: Resolver sch -> Term Wanted sch ty -> Term Maybe sch ty
```

Coupled with serialization from and to the Haskell values, we can get a nice interface which says:

```haskell
resolvers :: (ToSchema Wanted sch ty t, FromSchema Maybe sch ty r)
          => Resolver sch -> t -> r
```

### More complex execution

This design works, and would be the first goal for GraphQL. However, there are more things to consider:

- Resolvers may want to execute some actions, like querying a database. This means that `Resolver`s would need to look more like (notice the `m` in the field).

    ```haskell
    data FieldResolver (sch :: Schema) tyname m f where
      RField :: (Term sch (sch :/: tyname) -> m (FieldValue sch v))
             -> FieldResolver sch tyname m ('FieldDef fname v)
    ```

    I think that if `m` is an `Applicative`, this should pose no further complication.

    Using a monad there also generalizes one of the things done in Sangria, in which you can also carry some data or state throughout the resolver. In our case, that would mean `ReaderT` or `StateT`.

- We may want to allow more than one "round". Right now, the query would be traversed top-down, executing the resolver at each step. But what if we need to traverse the tree once, "decorate" the tree with some values, and then go top-down again? This is exactly the idea behind [attribute grammars](https://wiki.haskell.org/The_Monad.Reader/Issue4/Why_Attribute_Grammars_Matter)!

    In our case, that would mean that a "full `Resolver`" is actually a list of "chained simple `Resolvers`", each one of them taking as input the previous tree, and decorating something new. The "chaining" is visible in the change of functors:

    ```haskell
    data ResolverChain f g sch where
      Stop  :: ResolverChain f f sch
      (:#:) :: SimpleResolver f g sch
            -> ResolverChain g h sch
            -> ResolverChain f h sch
    
    type FullResolver = ResolverChain Wanted Maybe
    ```

    For more information about this design, you can check [*Reflection without Remorse*](https://dl.acm.org/citation.cfm?id=2633360), where it's used to speed up monadic sequences.

## Existing libraries

There is some work on GraphQL in the Haskell community. The most interesting one is `graphql`, because it defines [AST and parser](http://hackage.haskell.org/package/graphql-0.6.0.0/docs/Language-GraphQL-AST.html) for the query language. However, it's still quite limited, because it does not support subscriptions, and does not read GraphQL schema files. Thus, a first goal may be to make their parser support the missing pieces.

**Comment by serras** [comment]
It turns out there's a GraphQL library for Haskell called [Morpheus](https://morpheusgraphql.com/), which exposes an API similar to what I was proposing here. The only difference is (of course) that their types are not matched using our `Schema` data type.

**Comment by serras** [comment]
@cb372 has suggested [this video of GraphQL and OCaml](https://www.youtube.com/watch?v=jaKcEGkItsY). The design is very much as proposed here, except that in the video the definition of the schema and the resolvers and intetwined, whereas in this design the schema is taken from Mu.

**Comment by serras** [comment]
Closed by the new design in #111
